### PR TITLE
For #29218, Cancelling session renewal shouldn't lock you out of Shotgun

### PR DESCRIPTION
**Pull request by jfboismenu-adsk**
Implemented a simple wrapper method around renew_session (now renew_session_internal) that counts how many threads are renewing the session and clears the cancelled flag when the last one exits the method.

### CODE REVIEW

File: python/tank_vendor/shotgun_authentication/interactive_authentication.py [hunk 1]
```diff
@@ -366,18 +364,46 @@ def _renew_session(user, session_token, credentials_handler):
...
+
+# Makes access to the thread count and executing logic based on it thread
+# safe.
+_renew_session_lock = threading.Lock()
+# Number of threads who are trying to renew the session.
+_renew_session_thread_count = 0
+
+
+def _renew_session(user, session_token, credentials_handler):
+    # One more renewer.
```
Comment (manneohrstrom) [inline python/tank_vendor/shotgun_authentication/interactive_authentication.py]: please add doc string :)

File: python/tank_vendor/shotgun_authentication/interactive_authentication.py [hunk 2]
```diff
@@ -366,18 +364,46 @@ def _renew_session(user, session_token, credentials_handler):
...
+    global _renew_session_lock
+    with _renew_session_lock:
+        global _renew_session_thread_count
+        _renew_session_thread_count = _renew_session_thread_count + 1
```
Comment (manneohrstrom) [inline python/tank_vendor/shotgun_authentication/interactive_authentication.py]: FYI `_renew_session_thread_count += 1` works in python (but there is no ++ operator)

File: python/tank_vendor/shotgun_authentication/interactive_authentication.py (hunk 1 above)
Comment (manneohrstrom) [inline python/tank_vendor/shotgun_authentication/interactive_authentication.py]: would be great if docstring explains the use case we are trying to cater for (the multiple cancel clicking). Knowing the backstory will make it a lot easier to read and maintain the code.

**Comment by manneohrstrom** [comment]
Looking awesome! Just one comment about a docstring that I think could be useful for code readers.

File: python/tank_vendor/shotgun_authentication/interactive_authentication.py (hunk 1 above)
Comment (jfboismenu-adsk) [inline python/tank_vendor/shotgun_authentication/interactive_authentication.py]: That's so weird. I clearly remember adding that docstring originally. I'll add it again. Thanks!

Comment (jfboismenu-adsk) [inline python/tank_vendor/shotgun_authentication/interactive_authentication.py]: There it is! I knew I had documented it. Only I did it in a separate branch! https://github.com/shotgunsoftware/tk-core/pull/151/files#diff-6492fe4a8b359020014bf2944706aba7R413

**Pull request closed without merging**